internal/concierge/impersonator/impersonator.go

```diff
@@ -0,0 +1,153 @@
+// Copyright 2020-2021 the Pinniped contributors. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package impersonator
+
+import (
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"net/http/httputil"
+	"net/url"
+	"strings"
+
+	"github.com/go-logr/logr"
+	"k8s.io/apiserver/pkg/authentication/user"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/transport"
+
+	"go.pinniped.dev/generated/1.20/apis/concierge/login"
+	loginv1alpha1 "go.pinniped.dev/generated/1.20/apis/concierge/login/v1alpha1"
+	"go.pinniped.dev/internal/controller/authenticator/authncache"
+)
+
+// allowedHeaders are the set of HTTP headers that are allowed to be forwarded through the impersonation proxy.
+//nolint: gochecknoglobals
+var allowedHeaders = []string{
+	"Accept",
+	"Accept-Encoding",
+	"User-Agent",
+	"Connection",
+	"Upgrade",
+}
+
+type Proxy struct {
+	cache *authncache.Cache
+	proxy *httputil.ReverseProxy
+	log   logr.Logger
+}
+
+func New(cache *authncache.Cache, log logr.Logger) (*Proxy, error) {
+	return newInternal(cache, log, rest.InClusterConfig)
+}
+
+func newInternal(cache *authncache.Cache, log logr.Logger, getConfig func() (*rest.Config, error)) (*Proxy, error) {
+	kubeconfig, err := getConfig()
+	if err != nil {
+		return nil, fmt.Errorf("could not get in-cluster config: %w", err)
+	}
+
+	serverURL, err := url.Parse(kubeconfig.Host)
+	if err != nil {
+		return nil, fmt.Errorf("could not parse host URL from in-cluster config: %w", err)
+	}
+
+	kubeTransportConfig, err := kubeconfig.TransportConfig()
+	if err != nil {
+		return nil, fmt.Errorf("could not get in-cluster transport config: %w", err)
+	}
+	kubeTransportConfig.TLS.NextProtos = []string{"http/1.1"}
+
+	kubeRoundTripper, err := transport.New(kubeTransportConfig)
+	if err != nil {
+		return nil, fmt.Errorf("could not get in-cluster transport: %w", err)
+	}
+
+	proxy := httputil.NewSingleHostReverseProxy(serverURL)
+	proxy.Transport = kubeRoundTripper
+
+	return &Proxy{
+		cache: cache,
+		proxy: proxy,
+		log:   log,
+	}, nil
+}
+
+func (p *Proxy) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	log := p.log.WithValues(
+		"url", r.URL.String(),
+		"method", r.Method,
+	)
+
```

Comment:
This should fail if any impersonation headers are already set on the request.  Silently dropping them is not safe.

Reply:
Maybe we should hard fail on any unexpected headers?

Reply:
https://github.com/vmware-tanzu/pinniped/pull/355/commits/c7905c66386b229c1ac633fd4bae92e912cf76c1